README.md
# ResourceLink #

ResourceLink is an open source Python application to link resource files into
single one (*link* command). It also provides *map* command to generate a
resource map file in C array style, *bmap* command to generate the resource map
file in binary style, and *id* command to generate a C header file of resource
ID enumeration.


## Install ##

1. Download a binary distribution of ResourceLink (e.g.,
   *ResourceLink-1.14-bin.zip*) from [Downloads][] page.
2. Uncompress the binary distribution.

[Downloads]: https://bitbucket.org/YorkJong/pyresourcelink/downloads


## Getting Started ##

1. Install ResourceLink.
2. Put resource files to the *res* folder.
3. Edit the *res.lst* file.
4. Run `demo.bat` to generate *res.bin*, *ResMap.i*, *ResMap.bin*, and
   *ResID.h*.
    - *res.bin* is generated with **link** command.
    - *ResMap.i* is generated with **map** command.
    - *ResMap.bin* is generated with **bmap** command.
    - *ResID.h* is generated with **id** command.
5. Run `clean.bat` to remove generated files.

### The sample *res.lst* ###
```sh
:0x00       # start offset (default is 0x00)

:kind=B     # kind B for the enumeration
bat.png     # file size: 2877 bytes

:4096       # offset to address: 4096
broom.png   # file size: 3083 bytes

:kind=C1
:kind=C2

:kind=D
candle.png  # file size: 2771 bytes

:kind=E1
:kind=E2
```
- This file is used to list filenames of resources.
- A `#` denotes a comment begins after the `#` and ends at the end of the line.
- A line prefixing `:<number>` denotes an byte alignment of next resource when
  linking.
    - The <number> can be a decimal number or a hexadecimal number (begins
      with `0x`)
    - The `:4096` prefixing `broom.png` means *broom.png* will put on offset
      4096 bytes of the linking file.
- A line prefixing `:kind` works only on *id* command and is used to classify
  generated enumerators.
    - see the documentation of *generated ResID.h* for details

### generated *ResMap.i* ###
```c
// Generated by the Resource Link v1.14
//    !author: Jiang Yu-Kuan <yukuan.jiang@gmail.com>
//    !trail: reslnk.exe map -dres -oResMap.i -a4 res.lst

//   offset,       size     (in bytes)
{         0,       2877},   // RES_PNG_bat (bat.png)
{      4096,       3083},   // RES_PNG_broom (broom.png)
{      7180,       2771},   // RES_PNG_candle (candle.png)
```
* You can apply a generated *ResMap.i* with a C `struct` to read the content of
  a resource.

### An example C struc with *ResMap.i* looks like as follows: ###
```c
#include "ResID.h"


/** Resource Map (a list of offset-size pairs) */
static struct {
    uint32_t offset;    ///< offset of a resource
    uint32_t nBytes;    ///< total bytes of a resource
} _resMap[] = {
    #include "ResMap.i"
};


/** Returns address in NOR Flash that stores the resource. */
Addr Res_Addr(ResID id)
{
    assert (id < RES_End);

    Res_initMapTblIfHasNotDone();
    return _resMap[id].offset + _baseAddr;
}


/** Returns total bytes of a resource. */
uint32_t Res_Bytes(ResID id)
{
    assert (id < RES_End);

    Res_initMapTblIfHasNotDone();
    return _resMap[id].nBytes;
}
```


### generated *ResID.h* ###
```c
// Generated by the Resource Link v1.14
//    !author: Jiang Yu-Kuan <yukuan.jiang@gmail.com>
//    !trail: reslnk.exe id -oResID.h res.lst
#ifndef _RES_ID_H
#define _RES_ID_H


/** IDs of Resources */
typedef enum {
    RES_B_BEGIN,
    RES_PNG_bat = RES_B_BEGIN,
    RES_PNG_broom,
    RES_B_END,

    RES_C1_BEGIN = RES_B_END,
    RES_C1_END = RES_C1_BEGIN,

    RES_C2_BEGIN = RES_C1_END,
    RES_C2_END = RES_C2_BEGIN,

    RES_D_BEGIN = RES_C2_END,
    RES_PNG_candle = RES_D_BEGIN,
    RES_D_END,

    RES_E1_BEGIN = RES_D_END,
    RES_E1_END = RES_E1_BEGIN,

    RES_E2_BEGIN = RES_E1_END,
    RES_E2_END = RES_E2_BEGIN,

    RES_End = RES_E2_END,
    RES_Total = RES_End
} ResID;


#endif // _RES_ID_H
```
- The `RES_B_BEGIN` and `RES_B_END` are relative to the `:kind=B` in *res.lst*
  file.


## Command Line ##
### Top level ###
```
usage: reslnk.exe [-h] [-v] {link,map,bmap,id} ...

positional arguments:
  {link,map,bmap,id}
                        commands
    link                link resource files into single one.
    map                 generate a resource map file in format of C array.
    bmap                generate a resource map file in binary format.
    id                  generate a C header file of resource ID enumeration.

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
```

### link command ###
```
usage: reslnk.exe link [-h] [-d <directory>] [-a <number>] [-p <char hex>]
                       [-o <file>]
                       lst-file

positional arguments:
  lst-file              The list file of resources.

optional arguments:
  -h, --help            show this help message and exit
  -d <directory>, --dir <directory>
                        assign the <directory> to read resource files. The
                        default directory is ".".
  -a <number>, --align <number>
                        specify the <number> of alignment bytes (default "1").
  -p <char hex>, --padding <char hex>
                        specify the padding hex value of a char (default
                        "0xFF").
  -o <file>, --output <file>
                        place the output into <file>, the file after linking
                        (default "res.bin").
```

### map command ###
```
usage: reslnk.exe map [-h] [-d <directory>] [-a <number>] [-o <file>] lst-file

positional arguments:
  lst-file              The list file of resources.

optional arguments:
  -h, --help            show this help message and exit
  -d <directory>, --dir <directory>
                        assign the <directory> to read resource files. The
                        default directory is ".".
  -a <number>, --align <number>
                        specify the <number> of alignment bytes (default "1").
  -o <file>, --output <file>
                        place the output into <file>, the C included file
                        listing the offset, size pairs (default "ResMap.i").
```

### bmap command ###
```
usage: reslnk.exe bmap [-h] [-d <directory>] [-a <number>] [-o <file>] lst-file

positional arguments:
  lst-file              The list file of resources.

optional arguments:
  -h, --help            show this help message and exit
  -d <directory>, --dir <directory>
                        assign the <directory> to read resource files. The
                        default directory is ".".
  -a <number>, --align <number>
                        specify the <number> of alignment bytes (default "1").
  -o <file>, --output <file>
                        place the output into <file>, the binary version of
                        resource map file listing the offset, size pairs
                        (default "ResMap.bin").
```

### id command ###
```
usage: reslnk.exe id [-h] [-o <file>] lst-file

positional arguments:
  lst-file              The list file of resources.

optional arguments:
  -h, --help            show this help message and exit
  -o <file>, --output <file>
                        place the output into <file>, the C header file of
                        resource ID enumeration (default "ResID.h").
```



CHANGELOG.md
### 1.14 (2016-02-08)

- Extracted myutil.py from reslnk.py
- Added README.md
- Added CHANGELOG.md
- Added LICENSE.md
- Hosted to bitbucket.org
- Moved files for distribute.bat
- Removed filesize command
- Removed checksum command

### 1.13 (2015-04-07)

- Added allowing for *empty kinds* in generated ResID.h file

### 1.12 (2014-12-25)

- Added MD5 string (put at 0xA0) to the checksum command

### 1.11 (2014-11-20)
- Supported filenames with space characters

### 1.10 (2014-09-25)

- Added bmap command to generate binary formated map file.

### 1.00 (2013-08-22)

- Added the support of :kind command in resouce list file.

### 0.99 (2013-03-28)

- Added align option to map and link commands for specifying the number of
  alignment bytes.


### 0.12 (2013-03-22)

- Added usb_head command to generate USB ISP header file of A1016
- Renamed usb_head command to checksum command
- Added padding option to link command
- Added filesize command to generate a filesize header file
- Appended newline on generated resource map files (e.g. ResMap.i)

### 0.02 (2013-2-27)

- Initial version
- Added commands of link, map, and id



reslnk/myutil.py
# -*- coding: utf-8 -*-
"""
This module put my utility functions
"""
__author__ = "Jiang Yu-Kuan <yukuan.jiang@gmail.com>"
__date__ = "2016/02/08 (initial version) ~ 2016/02/08 (last revision)"

import re
import os
import sys


#------------------------------------------------------------------------------
# File
#------------------------------------------------------------------------------

def save_utf8_file(fn, lines):
    """Save string lines into an UTF8 text files.
    """
    out_file = open(fn, "w")
    out_file.write("\n".join(lines).encode("utf-8"))
    out_file.close()


def main_basename(path):
    """Return a main name of a basename of a given file path.

    Example
    -------
    >>> main_basename('c:\code\langconv\MsgID.h')
    'MsgID.h'
    """
    base = os.path.basename(path)
    base_main, base_ext = os.path.splitext(base)
    return base_main


#------------------------------------------------------------------------------
# Math
#------------------------------------------------------------------------------

def is_numeric(str):
    try:
        offset = int(eval(str))
    except:
        return False
    return True


#------------------------------------------------------------------------------
# String
#------------------------------------------------------------------------------

def replace_chars(text, replaced_pairs='', deleted_chars=''):
    """Return a char replaced text.

    Arguments
    ---------
    text -- the text
    replaced_pairs -- the replaced chars

    Example
    -------
    >>> replaced = [('a','b'), ('c','d')]
    >>> removed = 'e'
    >>> replace_chars('abcde', replaced, removed)
    'bbdd'
    """
    for old, new in replaced_pairs:
        text = text.replace(old, new)
    for ch in deleted_chars:
        text = text.replace(ch, '')
    return text


def camel_case(string):
    """Return camel case string from a space-separated string.

    Example
    -------
    >>> camel_case('good job')
    'GoodJob'
    """
    return ''.join(w.capitalize() for w in string.split())


def replace_punctuations(text):
    """Replace punctuation characters with abbreviations for a string.
    """
    punctuations = [
        ('?', 'Q'),   # Q:  question mark
        ('.', 'P'),   # P:  period; full stop
        ('!', 'E'),   # E:  exclamation mark
        ("'", 'SQ'),  # SQ: single quotation mark; single quote
        ('"', 'DQ'),  # DQ: double quotation mark; double quotes
        ('(', 'LP'),  # LP: left parenthese
        (')', 'RP'),  # RP: right parenthese
        (':', 'Cn'),  # Cn: colon
        (',', 'Ca'),  # Ca: comma
        (';', 'S'),   # S:  semicolon
    ]
    deleted = '+-*/^=%$#@|\\<>{}[]'
    return replace_chars(text, punctuations, deleted)


def remain_alnum(text):
    """Remain digits and English letters of a string.
    """
    return ''.join(c for c in text if c.isalnum()
                                   and ord(' ') <= ord(c) <= ord('z'))


#------------------------------------------------------------------------------
# For code generation
#------------------------------------------------------------------------------

def c_identifier(text):
    """Convert input text into an legal identifier in C.

    Example
    -------
    >>> c_identifier("Hello World")
    'HelloWorld'
    >>> c_identifier("Anti-Shake")
    'Antishake'
    """
    if ' ' in text:
        text = camel_case(text)
    text = re.sub(r'\+\d+', lambda x: x.group().replace('+', 'P'), text)
    text = re.sub(r'\-\d+', lambda x: x.group().replace('-', 'N'), text)
    text = replace_punctuations(text)
    return remain_alnum(text)


def wrap_header_guard(lines, h_fn):
    """Wrap a C header guard for a given line list.
    """
    def underscore(txt):
        """Return an under_scores text from a CamelCase text.

        This function will leave a CamelCase text unchanged.
        """
        s1 = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', txt)
        return re.sub('([a-z0-9])([A-Z])', r'\1_\2', s1).lower()

    h_fn_sig = '_%s_H' % underscore(main_basename(h_fn)).upper()
    begin = ['#ifndef %s' % h_fn_sig]
    begin += ['#define %s' % h_fn_sig, '', '']
    end = ['', '', '#endif // %s' % h_fn_sig, '']
    return begin + lines + end


def prefix_info(lines, software, version, author, comment_mark='//'):
    """Prefix information to the given lines with given comment-mark.
    """
    prefix = ['%s Generated by the %s v%s' % (comment_mark,
              software, version)]
    prefix += ['%s    !author: %s' % (comment_mark, author)]
    prefix += ['%s    !trail: %s %s' % (comment_mark,
               os.path.basename(sys.argv[0]), ' '.join(sys.argv[1:]))]
    return prefix + lines





reslnk/reslnk.py
# -*- coding: utf-8 -*-
"""
The main goal of ResLnk (Resource Link) is to link resource files into single
one (link command). It provides map command to generate resource map file in C
array style, id command to generate a C header file of resource ID enumeration.
It also provids additional commmands (e.g. checksum, and filesize) for the USB
boot and the bootloading on A1016 ICs.
"""
__software__ = "Resource Link"
__version__ = "1.14"
__author__ = "Jiang Yu-Kuan <yukuan.jiang@gmail.com>"
__date__ = "2013/02/26 (initial version) ~ 2016/02/08 (last revision)"

import os
import sys
import re
import argparse

from myutil import *


#-----------------------------------------------------------------------------
# File Read
#-----------------------------------------------------------------------------

def read_lst_file(fn):
    """Read a resource map file and return statement list.
    """
    def del_nonuse(lines):
        lines = (x.split("#", 1)[0].strip() for x in lines)
        return [x for x in lines if len(x) > 0]

    with open(fn) as f:
        lines = del_nonuse(f.read().splitlines())

    kind_cmd_pattern = re.compile(':\s*kind\s*=')
    for line in lines:
        if not line.startswith(':'):        # check filename
            continue
        if kind_cmd_pattern.match(line):    # check :kind command
            continue
        if is_numeric(line[1:]):            # check :offset
            continue
        raise ValueError('syntax error -> %s' % line)
    return lines


#-----------------------------------------------------------------------------
# Misc
#-----------------------------------------------------------------------------

def prefix_authorship(lines, comment_mark='//'):
    """Prefix authorship infomation to the given lines
    with given comment-mark.
    """
    return prefix_info(lines,
                       __software__, __version__, __author__,
                       comment_mark)


def res_id_from_filename(fn):
    """Return resource ID from filename.
    """
    base = os.path.basename(fn)
    base_main, base_ext = os.path.splitext(base)
    return 'RES_%s_%s' % (c_identifier(base_ext[1:]).upper(),
                          c_identifier(base_main.replace('_', ' ')))


def map_from_statements(statements, res_dir='res', align_bytes=1):
    """Return (offset, size, filename) tuples from statements.
    """
    assert align_bytes >= 1

    tuples = []
    offset = 0
    for sta in statements:
        if sta.startswith(':'):
            if not is_numeric(sta[1:]):
                continue
            val = eval(sta[1:])
            if val < offset:
                raise ValueError('offset too small -> %s' % sta)
            offset = val
        else:
            fn = '{path}/{name}'.format(path=res_dir, name=sta)
            fsize = os.path.getsize(fn)
            tuples += [(offset, fsize, sta)]
            offset += fsize
            offset = (offset + align_bytes - 1) / align_bytes * align_bytes
    return tuples


#-----------------------------------------------------------------------------
# Action Functions
#-----------------------------------------------------------------------------

def link(statements, res_dir='res', outfile='res.bin',
        align_bytes=1, padding_hex=0xFF):
    """Link resources into single file.
    """
    assert padding_hex <= 0xFF
    assert align_bytes >= 1

    begins, sizes, fns = zip(
        *map_from_statements(statements, res_dir, align_bytes))
    ends = list(begins[1:]) + [begins[-1] + sizes[-1]]
    spaces = [end - begin for begin, end in zip(begins, ends)]

    with open(outfile, 'wb') as outfile:
        for space, fn in zip(spaces, fns):
            fn = '{path}/{name}'.format(path=res_dir, name=fn)
            with open(fn, 'rb') as infile:
                raw = infile.read()
            padding = chr(padding_hex) * (space - len(raw))
            outfile.write(raw + padding)


def gen_map_ifile(statements,
                  res_dir='res', outfile='ResMap.i', align_bytes=1):
    """Generate a C included file that lists map of resources.
    """
    assert align_bytes >= 1

    lines = ['']
    lines += ['// %8s,   %8s     (in bytes)' % ('offset', 'size')]
    lines += ['{  %8d,   %8d},   // %s (%s)'
            % (offset, size, res_id_from_filename(fn), fn)
            for offset, size, fn
            in map_from_statements(statements, res_dir, align_bytes)]
    lines += ['']

    lines = prefix_authorship(lines)
    save_utf8_file(outfile, lines)


def gen_map_binfile(statements,
                    res_dir='res', outfile='ResMap.bin', align_bytes=1):
    """Generate a binary map file of resources.
    """
    def str_from_val(x):
        """"Return char-string from a two-byte value (high byte first)."""
        return (chr((x >> 24) & 0xff) +
                chr((x >> 16) & 0xff) +
                chr((x >> 8) & 0xff) +
                chr(x & 0xff))

    assert align_bytes >= 1

    lines = []
    for offset, size, fn in map_from_statements(statements, res_dir, align_bytes):
        lines += [str_from_val(offset), str_from_val(size)]
    bytes = 'ResMapTb' + ''.join(lines)

    with open(outfile, 'wb') as f:
        f.write(bytes)


def gen_id_hfile(statements, h_fn='ResID.h'):
    """Generate a C header file of resource ID enumeration.
    """
    def group_filenames(statements):
        def is_kept(sta):
            if sta.strip() == '':
                return False
            if not sta.startswith(':'):
                return True
            if re.match(':\s*kind\s*=', sta):
                return True
            return False

        def delete_space_lines(lines):
            return (line for line in lines if line != '')

        statements = [sta for sta in statements if is_kept(sta)]
        statements += ['']      # for last empty kind statement
        lines = '\n'.join(statements)
        kinds = re.findall(r':\s*kind\s*=\s*(.*)\s*\n', lines)
        fns_lst = re.split(r':\s*kind\s*=.*\n', lines)
        if statements[0].startswith(':'):
            fns_lst = fns_lst[1:]
        else:
            kinds = [''] + kinds
        fns_lst = (delete_space_lines(lines.split('\n')) for lines in fns_lst)
        return zip(kinds, fns_lst)

    def lines_from_fns(kind, fns, id_end):
        ids = [res_id_from_filename(fn) for fn in fns]
        id_begin = res_id_from_filename('BEGIN.' + kind)

        lines = []
        if id_end == '':
            lines += ['    %s,' % id_begin]
        else:
            lines += ['    %s = %s,' % (id_begin, id_end)]
        id_end = res_id_from_filename('END.' + kind)
        if ids:
            lines += ['    %s = %s,' % (ids[0], id_begin)]
            lines += ['    %s,' % id for id in ids[1:]]
            lines += ['    %s,' % id_end]
        else:
            lines += ['    %s = %s,' % (id_end, id_begin)]
        lines += ['']
        return lines, id_end

    lines = ['/** IDs of Resources */']
    lines += ["typedef enum {"]

    id_end = ''
    for kind, fns in group_filenames(statements):
        tmp_lines, id_end = lines_from_fns(kind, fns, id_end)
        lines += tmp_lines

    lines += ['    RES_End = %s,' % id_end]
    lines += ['    RES_Total = RES_End']
    lines += ['} ResID;']

    lines = wrap_header_guard(lines, h_fn)
    lines = prefix_authorship(lines, comment_mark='//')
    save_utf8_file(h_fn, lines)


#-----------------------------------------------------------------------------
# Command Line Interface
#-----------------------------------------------------------------------------

def parse_args(args):
    def do_link(args):
        link(args.statements, args.dir, args.outfile, args.align, args.padding)

    def do_map(args):
        gen_map_ifile(args.statements, args.dir, args.outfile, args.align)

    def do_bmap(args):
        gen_map_binfile(args.statements, args.dir, args.outfile, args.align)

    def do_id(args):
        gen_id_hfile(args.statements, args.outfile)

    def hex_check(value):
        value = int(eval(value))
        assert value < 256
        return value

    # create top-level parser
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('-v', '--version', action='version',
                        version='%s v%s by %s' %
                        (__software__, __version__, __author__))
    subparsers = parser.add_subparsers(help='commands')

    #--------------------------------------------------------------------------

    # create the parent parser of resource map file
    src = argparse.ArgumentParser(add_help=False)
    src.add_argument('statements', metavar='lst-file',
        type=read_lst_file,
        help='The list file of resources.')

    # create the parent parser of resource directory
    dir = argparse.ArgumentParser(add_help=False)
    dir.set_defaults(dir='.')
    dir.add_argument('-d', '--dir', metavar='<directory>',
        help="""assign the <directory> to read resource files.
            The default directory is "%s".
            """ % dir.get_default('dir'))

    # create the parent parser of align bytes
    aln = argparse.ArgumentParser(add_help=False)
    aln.set_defaults(align=1)
    aln.add_argument('-a', '--align', metavar='<number>',
        type=int, choices=(1, 2, 4, 8, 16, 32, 64, 128, 256),
        help="""specify the <number> of alignment bytes
            (default "%d").
            """ % aln.get_default('align'))

    # create the parser for the "link" command
    sub = subparsers.add_parser('link', parents=[src, dir, aln],
        help='link resource files into single one.')
    sub.set_defaults(func=do_link,
        outfile='res.bin', padding=0xFF)
    sub.add_argument('-p', '--padding', metavar='<char hex>', type=hex_check,
        help='''specify the padding hex value of a char (default "0x%X").
            ''' % sub.get_default('padding'))
    sub.add_argument('-o', '--output', metavar='<file>', dest='outfile',
        help='''place the output into <file>, the file after linking
            (default "%s").
            ''' % sub.get_default('outfile'))

    # create the parser for the "map" command
    sub = subparsers.add_parser('map', parents=[src, dir, aln],
        help='generate a resource map file in format of C array.')
    sub.set_defaults(func=do_map,
        outfile='ResMap.i')
    sub.add_argument('-o', '--output', metavar='<file>', dest='outfile',
        help='''place the output into <file>, the C included file listing
            the offset, size pairs (default "%s").
            ''' % sub.get_default('outfile'))

    # create the parser for the "bmap" command
    sub = subparsers.add_parser('bmap', parents=[src, dir, aln],
        help='generate a resource map file in binary format.')
    sub.set_defaults(func=do_bmap,
        outfile='ResMap.bin')
    sub.add_argument('-o', '--output', metavar='<file>', dest='outfile',
        help='''place the output into <file>, the binary version of resource
            map file listing the offset, size pairs (default "%s").
            ''' % sub.get_default('outfile'))

    # create the parser for the "id" command
    sub = subparsers.add_parser('id', parents=[src],
        help='generate a C header file of resource ID enumeration.')
    sub.set_defaults(func=do_id,
        outfile='ResID.h')
    sub.add_argument('-o', '--output', metavar='<file>', dest='outfile',
        help='''place the output into <file>, the C header file of
            resource ID enumeration (default "%s").
            ''' % sub.get_default('outfile'))

    #--------------------------------------------------------------------------

    # parse args and execute functions
    args = parser.parse_args(args)
    args.func(args)


def main():
    """Start point of this module.
    """
    try:
        parse_args(sys.argv[1:])
    except IOError as err:
        print err
    except ValueError as err:
        print err
    except KeyError as key:
        print 'KeyError:', key


if __name__ == '__main__':
    main()
    #parse_args(sys.argv[1:])

